plutus-use-cases/test/Spec/Marlowe.hs

```diff
@@ -0,0 +1,353 @@
+        testProperty "Oracle Commit/Pay works" oraclePayment,
+        testProperty "Escrow Contract" escrowTest,
+        testProperty "Futures" futuresTest,
+        testProperty "invalid contract: duplicate IdentCC" duplicateIdentCC,
+        testProperty "can't commit after timeout" cantCommitAfterStartTimeout,
+        testProperty "redeem after commit expired" redeemAfterCommitExpired
+        ]
+
+-- | Run a trace with the given scenario and check that the emulator finished
+--   successfully with an empty transaction pool.
+checkMarloweTrace :: MarloweScenario -> Trace MockWallet () -> Property
+checkMarloweTrace MarloweScenario{mlInitialBalances} t = property $ do
+    let model = Gen.generatorModel { Gen.gmInitialBalance = mlInitialBalances }
+    (result, st) <- forAll $ Gen.runTraceOn model t
+    Hedgehog.assert (isRight result)
+    Hedgehog.assert ([] == _txPool st)
+
+
+updateAll :: [Wallet] -> Trace MockWallet ()
+updateAll wallets = processPending >>= void . walletsNotifyBlock wallets
+
+getScriptOutFromTx :: Tx -> (TxOut', TxOutRef')
+getScriptOutFromTx tx = head . filter (isPayToScriptOut . fst) . txOutRefs $ tx
+
+performs :: Wallet -> m () -> Trace m (TxOut', TxOutRef')
+performs actor action = do
+    [tx] <- walletAction actor action
+    processPending >>= void . walletsNotifyBlock [actor]
+    assertIsValidated tx
+    return $ getScriptOutFromTx tx
+
+withContract
+    :: [Wallet]
+    -> Contract
+    -> ((TxOut', TxOutRef') -> Trace MockWallet ((TxOut', TxOutRef'), State))
+    -> Trace MockWallet ()
+withContract wallets contract f = do
+    [tx] <- walletAction creator (createContract contract 12)
+    let txOut = getScriptOutFromTx tx
+    update
+    assertIsValidated tx
+
+    (tx1Out, state) <- f txOut
+
+    [tx] <- walletAction creator (endContract tx1Out state)
+    update
+    assertIsValidated tx
+  where
+    creator = head wallets
+    update  = updateAll wallets
+
+oraclePayment :: Property
+oraclePayment = checkMarloweTrace (MarloweScenario {
+    mlInitialBalances = Map.fromList [ (PubKey 1, 1000), (PubKey 2, 777) ] }) $ do
+    -- Init a contract
+    let alice = Wallet 1
+        bob = Wallet 2
+        oracle = PubKey 42
+        update = updateAll [alice, bob]
+    update
+
+    let contract = CommitCash (IdentCC 1) (PubKey 2) (ValueFromOracle oracle (Value 0)) 128 256
+            (Pay (IdentPay 1) (PubKey 2) (PubKey 1) (Committed (IdentCC 1)) 256 Null)
+            Null
+
+    let oracleValue = OracleValue oracle (Height 2) 100
+
+    withContract [alice, bob] contract $ \txOut -> do
+        txOut <- bob `performs` commit
+            txOut
+            [oracleValue] []
+            (IdentCC 1)
+            100
+            (State [(IdentCC 1, (PubKey 2, NotRedeemed 100 256))] [])
+            (Pay (IdentPay 1) (PubKey 2) (PubKey 1) (Committed (IdentCC 1)) 256 Null)
+
+        txOut <- alice `performs` receivePayment txOut
```

Comment:
I wonder how this does not make `-Wall` freak out. So many different things called `txOut`. Could you use something like `txOut0`, `txOut1`, `txOut2` at least? Or even better: `txOutInit`, `txOutAfterBob`, `txOutAfterAlice`.